package.json

```
@@ -3,7 +3,29 @@
   "title": "jQuery",
   "description": "JavaScript library for DOM operations",
   "version": "4.0.0-pre",
+  "type": "commonjs",
+  "exports": {
+    ".": {
+      "node": {
+        "module": "./dist/jquery.mjs",
+        "import": "./dist/jquery-node-wrapper.mjs",
+        "default": "./dist/jquery.js"
+      },
+      "browser": "./dist/jquery.mjs"
```

You probably want to add the `require` and `default` in case, that should cover most cases. I'm not sure if `"browser"` is needed if you already have `default`, but also can't hurt.
```suggestion
      "require": "./dist/jquery.js",
      "browser": "./dist/jquery.mjs",
      "default": "./dist/jquery.mjs"
```

---

I would also recommend flipping around the file extensions. Use `.cjs` for the CommonJS version and `.js` for all other versions, and add `"type": "module"` to the `package.json`. I think it’s still the case that many servers/CDNs don’t serve the proper MIME type for `.mjs` files, so the browser version of jQuery (even the browser ESM version) should absolutely have a `.js` file extension for maximum compatibility.

---

> I think it’s still the case that many servers/CDNs don’t serve the proper MIME type for `.mjs` files

It's been years since I've since this issue, not sure if it's worth it. Note that the `.js` files are not CJS but AMD FWIW.

---

> It’s been years since I’ve since this issue, not sure if it’s worth it.

I think jQuery’s enormous scope justifies being careful. There’s no _benefit_ to `.mjs`/`.js` over `.cjs`/`.js`. Node 10 doesn’t support `"exports"` at all; if you want to explicitly provide a Node 10-and-below ESM build, an `.mjs` file somewhere just for those consumers could be included. I would still save all the others as `.js` files as `.js` works everywhere.

---

@GeoffreyBooth I'd like to understand where every entry comes from. Why do we need `require`, `browser` & `default` for non-Node.js cases?

---

> Why do we need `require` […] for non-Node.js cases?

E.g. if someone uses `require()` to write browser focused code. If you set `browser` to point to the ESM file, depending on the builder it may or may not work (i.e. it would work only if the bundler supports requiring ES modules).